core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala

```diff
@@ -625,6 +636,10 @@ final class ShuffleBlockFetcherIterator(
 
       result match {
         case r @ SuccessFetchResult(blockId, mapIndex, address, size, buf, isNetworkReqDone) =>
+          failureFetchResult.foreach { ffr =>
+            buf.release()
+            throwFetchFailedException(ffr.blockId, ffr.mapIndex, ffr.address, ffr.e)
```

Comment:
errr...I just feel that this makes things be more complex... I actually think that *fail-fast* here is not such necessary unless we really have high delay in current way.